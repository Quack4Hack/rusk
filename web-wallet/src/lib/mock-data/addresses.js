export default [
	"2087290d3dc213d43e493f03f5435f99",
	"ffbee869aca5ff5ee13c2706e5d9779d",
	"06527a34e1c91fc5785ea7764a0c34b1",
	"f62d307103ca54516b29fcedd5463d16"
];
