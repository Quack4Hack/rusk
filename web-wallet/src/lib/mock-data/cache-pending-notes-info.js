/** @type {WalletCachePendingNoteInfo[]} */
export default [
  {
    nullifier: Uint8Array.from([
      90, 206, 164, 251, 223, 185, 213, 67, 253, 217, 249, 241, 120, 255, 105,
      64, 110, 71, 47, 157, 193, 140, 35, 133, 69, 67, 228, 34, 223, 125, 236,
      46,
    ]),
    txId: "9bf1a3eac7fa0b8292c9351f0933df56e9468fdb30e79e145b2699b20778d42f",
  },
  {
    nullifier: Uint8Array.from([
      245, 135, 186, 46, 3, 142, 188, 44, 29, 178, 128, 202, 68, 53, 117, 62,
      10, 113, 73, 173, 248, 198, 106, 150, 160, 74, 86, 216, 87, 187, 59, 197,
    ]),
    txId: "1011e348e45fc3bb09f6a43888e51a1b71f8dee02755486f8948026951dc052d",
  },
  {
    nullifier: Uint8Array.from([
      126, 1, 67, 106, 78, 193, 10, 9, 193, 99, 2, 7, 107, 141, 64, 213, 223, 1,
      22, 187, 118, 9, 144, 128, 23, 122, 177, 197, 162, 141, 233, 145,
    ]),
    txId: "2b7db6a7c6ebc0b6d410971e9df83854ff11a39d9ba95fd2fbd84378343f436f",
  },
];
