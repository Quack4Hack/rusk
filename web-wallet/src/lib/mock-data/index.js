export { default as addresses } from "./addresses";
export { default as apiMarketData } from "./api-market-data.json";
export { default as stakeInfo } from "./stakeInfo";
export { default as transactions } from "./transactions";
