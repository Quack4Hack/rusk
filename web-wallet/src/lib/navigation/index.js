export { default as logout } from "./logout";
export { default as addBasePath } from "./addBasePath";
export { default as goto } from "./goto";
export { default as redirect } from "./redirect";
