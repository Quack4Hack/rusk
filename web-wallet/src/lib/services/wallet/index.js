export { default as getWallet } from "./getWallet";
