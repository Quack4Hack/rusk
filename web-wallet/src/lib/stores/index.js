export { default as operationsStore } from "./operationsStore";
export { default as settingsStore } from "./settingsStore";
export { default as walletStore } from "./walletStore";
