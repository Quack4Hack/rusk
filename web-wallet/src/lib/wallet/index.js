export { default as decryptMnemonic } from "./decryptMnemonic";
export { default as encryptMnemonic } from "./encryptMnemonic";
export { default as getSeedFromMnemonic } from "./getSeedFromMnemonic";
export { default as initializeWallet } from "./initializeWallet";
export { default as refreshLocalStoragePasswordInfo } from "./refreshLocalStoragePasswordInfo";
