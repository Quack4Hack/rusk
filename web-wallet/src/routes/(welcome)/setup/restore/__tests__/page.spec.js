import {
	afterAll,
	afterEach,
	describe,
	expect,
	it,
	vi
} from "vitest";
import { cleanup, fireEvent, render } from "@testing-library/svelte";
import { generateMnemonic } from "bip39";
import { Wallet } from "@dusk-network/dusk-wallet-js";
import { setKey } from "lamb";
import { get } from "svelte/store";
import { addresses } from "$lib/mock-data";
import * as navigation from "$lib/navigation";
import { settingsStore, walletStore } from "$lib/stores";
import { encryptMnemonic, getSeedFromMnemonic } from "$lib/wallet";
import loginInfoStorage from "$lib/services/loginInfoStorage";
import * as walletService from "$lib/services/wallet";
import { toastList } from "$lib/dusk/components/Toast/store";
import Restore from "../+page.svelte";
import { tick } from "svelte";

/**
 * @param {HTMLElement} input
 * @param {*} value
 * @returns {Promise<boolean>}
 */
const fireInput = (input, value) => fireEvent.input(input, { target: { value } });

/** @param {HTMLElement} element */
function asInput (element) {
	// eslint-disable-next-line no-extra-parens
	return /** @type {HTMLInputElement} */ (element);
}

describe("Restore", async () => {
	const walletGetPsksSpy = vi.spyOn(Wallet.prototype, "getPsks").mockResolvedValue(addresses);
	const mnemonic = generateMnemonic();
	const invalidMnemonic = "dad dad dad dad dad dad dad dad dad dad dad dad";
	const pwd = "passwordpassword";
	const loginInfo = await encryptMnemonic(mnemonic, pwd);
	const seed = getSeedFromMnemonic(mnemonic);
	const userId = (await new Wallet(seed).getPsks())[0];
	const getWalletSpy = vi.spyOn(walletService, "getWallet");
	const gotoSpy = vi.spyOn(navigation, "goto");
	const settingsResetSpy = vi.spyOn(settingsStore, "reset");
	const clearAndInitSpy = vi.spyOn(walletStore, "clearLocalDataAndInit");
	const readTextMock = vi.fn().mockResolvedValue(mnemonic);

	Object.assign(window.navigator, { clipboard: { readText: readTextMock } });

	afterEach(async () => {
		cleanup();
		clearAndInitSpy.mockClear();
		getWalletSpy.mockClear();
		gotoSpy.mockClear();
		settingsStore.reset();
		settingsResetSpy.mockClear();
		walletGetPsksSpy.mockClear();
		walletStore.reset();
		readTextMock.mockClear();
	});

	afterAll(() => {
		clearAndInitSpy.mockRestore();
		getWalletSpy.mockRestore();
		gotoSpy.mockRestore();
		settingsResetSpy.mockRestore();
		walletGetPsksSpy.mockRestore();
	});

	it("should render the Existing Wallet notice step of the Restore flow if there is a userId saved in localStorage", () => {
		settingsStore.update(setKey("userId", userId));

		const { container } = render(Restore);

		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render the Terms of Service step of the Restore flow if there is no userId saved in localStorage", () => {
		const { container } = render(Restore);

		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render the Mnemonic Authenticate step after accepting the Existing Wallet Notice and the Terms of Service", async () => {
		settingsStore.update(setKey("userId", userId));

		const { container, getByRole } = render(Restore);

		await fireEvent.click(getByRole("button", { name: "Proceed" }));
		await fireEvent.click(getByRole("button", { name: "Accept" }));

		expect(container.firstChild).toMatchSnapshot();
	});

	it("should allow the user to proceed to password setup after a valid mnemonic has been provided", async () => {
		const { getByRole } = render(Restore);

		await fireEvent.click(getByRole("button", { name: "Accept" }));

		const nextButton = getByRole("button", { name: "Next" });

		expect(nextButton).toBeDisabled();

		await fireEvent.click(getByRole("button", { name: "Paste seed phrase" }));

		await tick();
		expect(nextButton).not.toBeDisabled();
	});

	it("should not allow the user to proceed to password setup after an invalid mnemonic has been provided", async () => {
		readTextMock.mockResolvedValueOnce(invalidMnemonic);

		const { getByRole } = render(Restore);

		await fireEvent.click(getByRole("button", { name: "Accept" }));

		const nextButton = getByRole("button", { name: "Next" });

		expect(nextButton).toBeDisabled();

		expect(get(toastList).length).toBe(0);

		await fireEvent.click(getByRole("button", { name: "Paste seed phrase" }));
		await tick();

		expect(get(toastList).length).toBe(1);
		expect(nextButton).toBeDisabled();
	});

	it("should initialize the wallet without setting a password", async () => {
		loginInfoStorage.set(loginInfo);

		const { getByRole } = render(Restore);

		// ToS step
		await fireEvent.click(getByRole("button", { name: "Accept" }));

		// Mnemonic Authenticate step
		const nextButton = getByRole("button", { name: "Next" });

		expect(nextButton).toBeDisabled();

		await fireEvent.click(getByRole("button", { name: "Paste seed phrase" }));
		await tick();
		expect(nextButton).toBeEnabled();
		await fireEvent.click(nextButton);

		// Set Password step
		await fireEvent.click(getByRole("button", { name: "Next" }));
		expect(loginInfoStorage.get()).toBeNull();

		// Restore Wallet step
		await fireEvent.click(getByRole("button", { name: "Next" }));

		await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

		expect(settingsResetSpy).toHaveBeenCalledTimes(1);
		expect(getWalletSpy).toHaveBeenCalledTimes(1);
		expect(getWalletSpy).toHaveBeenCalledWith(seed);
		expect(clearAndInitSpy).toHaveBeenCalledTimes(1);
		expect(clearAndInitSpy).toHaveBeenCalledWith(expect.any(Wallet));
		expect(gotoSpy).toHaveBeenCalledTimes(1);
		expect(gotoSpy).toHaveBeenCalledWith("/dashboard");
	});

	it("should initialize the wallet with the encrypted mnemonic saved in localStorage", async () => {
		const { getByRole, getByPlaceholderText } = render(Restore);

		// ToS step
		await fireEvent.click(getByRole("button", { name: "Accept" }));

		// Mnemonic Authenticate step
		const nextButton = getByRole("button", { name: "Next" });

		expect(nextButton).toBeDisabled();

		await fireEvent.click(getByRole("button", { name: "Paste seed phrase" }));
		await tick();
		expect(nextButton).toBeEnabled();
		await fireEvent.click(nextButton);

		// Set Password step
		expect(loginInfoStorage.get()).toBeNull();

		await fireEvent.click(getByRole("switch"));

		await fireInput(asInput(getByPlaceholderText("Set Password")), pwd);
		await fireInput(asInput(getByPlaceholderText("Confirm Password")), pwd);

		expect(loginInfoStorage.get()).toBeNull();

		await fireEvent.click(getByRole("button", { name: "Next" }));
		await vi.waitFor(() => {
			expect(loginInfoStorage.get()).not.toBeNull();
		});

		// Restore Wallet step
		await fireEvent.click(getByRole("button", { name: "Next" }));
		await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

		expect(settingsResetSpy).toHaveBeenCalledTimes(1);
		expect(getWalletSpy).toHaveBeenCalledTimes(1);
		expect(getWalletSpy).toHaveBeenCalledWith(seed);
		expect(clearAndInitSpy).toHaveBeenCalledTimes(1);
		expect(clearAndInitSpy).toHaveBeenCalledWith(expect.any(Wallet));
		expect(gotoSpy).toHaveBeenCalledTimes(1);
		expect(gotoSpy).toHaveBeenCalledWith("/dashboard");
	});
});
