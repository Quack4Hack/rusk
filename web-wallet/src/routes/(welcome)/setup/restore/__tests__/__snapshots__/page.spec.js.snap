// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Restore > should render the Existing Wallet notice step of the Restore flow if there is a userId saved in localStorage 1`] = `
<div
  class="onboarding-wrapper svelte-i0tbsw"
>
  <section>
    <div
      class="dusk-card dusk-card--gap-default"
    >
      <div
        class="dusk-card__header-container"
      >
        <header
          class="dusk-card__header"
          slot="header"
        >
          <div
            class="dusk-card__heading-container dusk-card__heading-container--warning-card"
          >
            <svg
              class="dusk-icon dusk-icon--size--default"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M12,2L1,21H23M12,6L19.53,19H4.47M11,10V14H13V10M11,16V18H13V16"
              />
            </svg>
            
             
            <h3
              class="h4"
            >
              Existing Wallet Detected
            </h3>
          </div>
        </header>
      </div>
       
      <div
        class="dusk-card__body-container"
      >
        <p>
          Initializing a new wallet will replace your existing local wallet cache,
      erasing any stored data. Ensure you have securely backed up your current
      wallet's mnemonic phrase to prevent loss. Proceeding without a backup can
      lead to irreversible loss of access to your assets.
        </p>
         
        <div
          class="actions svelte-3n8m2w"
        >
          <a
            aria-disabled="false"
            class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--default actions__action"
            href="/some-base-path/setup"
          >
            <span
              class="dusk-anchor-button__text"
            >
              Back
            </span>
            
          </a>
           
          <button
            class="dusk-button dusk-button--type--button dusk-button--variant--primary dusk-button--size--default actions__action"
            type="button"
          >
            <span
              class="dusk-button__text"
            >
              Proceed
            </span>
          </button>
        </div>
      </div>
       
    </div>
    
  </section>
</div>
`;

exports[`Restore > should render the Mnemonic Authenticate step after accepting the Existing Wallet Notice and the Terms of Service 1`] = `
<div
  class="dusk-wizard dusk-wizard--full-height"
>
   
  <h2
    class="h1"
    slot="heading"
  >
    Enter
    <br />
     
    <mark>
      Mnemonic Phrase
    </mark>
  </h2>
   
  <div
    class="dusk-stepper dusk-stepper--variant--secondary"
    style="--columns: 5; --progress-width: 0%;"
  >
    <span
      aria-current="step"
      class="dusk-stepper__step dusk-stepper__step--processed"
    >
      1
    </span>
    <span
      class="dusk-stepper__step"
    >
      2
    </span>
    <span
      class="dusk-stepper__step"
    >
      3
    </span>
    <span
      class="dusk-stepper__step"
    >
      4
    </span>
    <span
      class="dusk-stepper__step"
    >
      5
    </span>
    
  </div>
  
   
  <div
    class="dusk-card dusk-card--gap-medium"
  >
    <div
      class="dusk-card__header-container"
    >
      <header
        class="dusk-card__header"
        slot="header"
      >
        <div
          class="dusk-card__heading-container"
        >
          <h3
            class="h4"
          >
            Enter your Mnemonic Phrase
          </h3>
           
        </div>
      </header>
    </div>
     
    <div
      class="dusk-card__body-container"
    >
      <div
        class="dusk-mnemonic"
      >
        <div
          class="dusk-mnemonic__actions-wrapper"
        >
          <button
            class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default dusk-icon-button--labeled"
            type="button"
          >
            <svg
              class="dusk-icon dusk-icon--size--default dusk-button__icon"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M19,20H5V4H7V7H17V4H19M12,2A1,1 0 0,1 13,3A1,1 0 0,1 12,4A1,1 0 0,1 11,3A1,1 0 0,1 12,2M19,2H14.82C14.4,0.84 13.3,0 12,0C10.7,0 9.6,0.84 9.18,2H5A2,2 0 0,0 3,4V20A2,2 0 0,0 5,22H19A2,2 0 0,0 21,20V4A2,2 0 0,0 19,2Z"
              />
            </svg>
            
             
            <span
              class="dusk-button__text"
            >
              Paste mnemonic phrase
            </span>
            
          </button>
           
          <button
            class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default dusk-icon-button--labeled"
            disabled=""
            type="button"
          >
            <svg
              class="dusk-icon dusk-icon--size--default dusk-button__icon"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M10.5,7A6.5,6.5 0 0,0 4,13.5A6.5,6.5 0 0,0 10.5,20H14V18H10.5C8,18 6,16 6,13.5C6,11 8,9 10.5,9H16.17L13.09,12.09L14.5,13.5L20,8L14.5,2.5L13.08,3.91L16.17,7H10.5M18,18H16V20H18V18Z"
              />
            </svg>
            
             
            <span
              class="dusk-button__text"
            >
              Undo
            </span>
            
          </button>
        </div>
         
        <ol
          class="dusk-words"
        >
          <li
            class="dusk-words__word"
          >
            _____
          </li>
          <li
            class="dusk-words__word"
          >
            _____
          </li>
          <li
            class="dusk-words__word"
          >
            _____
          </li>
          <li
            class="dusk-words__word"
          >
            _____
          </li>
          <li
            class="dusk-words__word"
          >
            _____
          </li>
          <li
            class="dusk-words__word"
          >
            _____
          </li>
          <li
            class="dusk-words__word"
          >
            _____
          </li>
          <li
            class="dusk-words__word"
          >
            _____
          </li>
          <li
            class="dusk-words__word"
          >
            _____
          </li>
          <li
            class="dusk-words__word"
          >
            _____
          </li>
          <li
            class="dusk-words__word"
          >
            _____
          </li>
          <li
            class="dusk-words__word"
          >
            _____
          </li>
        </ol>
         
        <div
          class="dusk-mnemonic__authenticate-actions-wrapper"
        >
          <input
            class="dusk-textbox dusk-textbox-text"
            maxlength="8"
            placeholder="Enter word 1"
            type="text"
          />
          
           
          
        </div>
      </div>
    </div>
     
  </div>
  
   
  <div
    class="dusk-wizard__step-navigation"
  >
    <a
      aria-disabled="false"
      class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--default dusk-icon-button--labeled"
      href="/some-base-path/setup"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-anchor-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
        />
      </svg>
      
       
      <span
        class="dusk-anchor-button__text"
      >
        Back
      </span>
      
      
    </a>
    
     
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default dusk-icon-button--labeled"
      disabled=""
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        Next
      </span>
       
      <svg
        class="dusk-icon dusk-icon--size--default dusk-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M4,11V13H16L10.5,18.5L11.92,19.92L19.84,12L11.92,4.08L10.5,5.5L16,11H4Z"
        />
      </svg>
      
    </button>
  </div>
  
  
   
  
   
  
   
  
   
  
</div>
`;

exports[`Restore > should render the Terms of Service step of the Restore flow if there is no userId saved in localStorage 1`] = `
<div
  class="onboarding-wrapper svelte-i0tbsw"
>
  <section>
    <div
      class="dusk-card dusk-card--gap-default"
    >
      <div
        class="dusk-card__header-container"
      >
        <header
          class="dusk-card__header"
          slot="header"
        >
          <div
            class="dusk-card__heading-container"
          >
            <h3
              class="h4"
            >
              Terms & Privacy Policy
            </h3>
             
          </div>
        </header>
      </div>
       
      <div
        class="dusk-card__body-container"
      >
        <p>
          Our 
          <a
            class="dusk-anchor dusk-anchor--on-surface"
            href="https://dusk.network/privacy-policy/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Terms & Privacy Policy
          </a>
           govern your use of our services, including data handling and user responsibilities.
      Your privacy and security are our top priorities.
        </p>
         
        <div
          class="actions svelte-3n8m2w"
        >
          <a
            aria-disabled="false"
            class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--default actions__action"
            href="/some-base-path/setup"
          >
            <span
              class="dusk-anchor-button__text"
            >
              Cancel
            </span>
            
          </a>
           
          <button
            class="dusk-button dusk-button--type--button dusk-button--variant--primary dusk-button--size--default actions__action"
            type="button"
          >
            <span
              class="dusk-button__text"
            >
              Accept
            </span>
          </button>
        </div>
      </div>
       
    </div>
    
  </section>
</div>
`;
